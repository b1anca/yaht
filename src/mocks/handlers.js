import { rest } from "msw";

export const handlers = [
  rest.post(
    `${process.env.REACT_APP_YAHT_API_URL}/auth/login`,
    (req, res, ctx) => {
      const { email } = req.body;

      return res(
        ctx.json({ token: `token-${email}`, name: "user name", email }),
        ctx.status(200)
      );
    }
  ),
];
