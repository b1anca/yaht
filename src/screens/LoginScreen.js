import React, { useEffect } from "react";
import { useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { userLogin } from "../features/auth/authActions";
import Error from "../components/Error";
import Input from "../components/Input";
import Button from "../components/Button";
import NavLink from "../components/NavLink";

function LoginScreen() {
  const { loading, userInfo, error } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const { register, handleSubmit } = useForm();

  const navigate = useNavigate();

  useEffect(() => {
    if (userInfo) {
      navigate("/dashboard");
    }
  }, [navigate, userInfo]);

  const submitForm = (data) => {
    dispatch(userLogin(data));
  };

  return (
    <main className="relative h-screen flex flex-1 flex-col overflow-hidden px-4 py-8 sm:px-6 lg:px-8">
      <div className="relative flex flex-1 flex-col items-center justify-center pb-16 pt-12">
        <form className="w-full max-w-sm" onSubmit={handleSubmit(submitForm)}>
          {error && <Error>{error}</Error>}
          <Input type="email" required {...register("email")} label="Email" />
          <Input
            type="password"
            required
            {...register("password")}
            label="Password"
          />
          <Button primary type="submit" loading={loading}>
            Sign in
          </Button>
        </form>
      </div>
      <footer className="relative shrink-0">
        <div className="space-y-4 text-sm text-gray-700 sm:flex sm:items-center sm:justify-center sm:space-x-4 sm:space-y-0">
          <p className="text-center sm:text-left">
            Don&apos;t have an account?
          </p>
          <NavLink secondary to="/register">
            Register
          </NavLink>
        </div>
      </footer>
    </main>
  );
}

export default LoginScreen;
