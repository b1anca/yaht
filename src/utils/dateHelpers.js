export const getDaysInMonth = (month, year) => {
  const date = new Date(year, month, 1);
  const days = [];
  while (date.getMonth() === month) {
    days.push(new Date(date));
    date.setDate(date.getDate() + 1);
  }
  return days;
};

export const formatDate = (date) => {
  const monthNames = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
  ];
  const day = date.getDate();
  const monthIndex = date.getMonth();
  const year = date.getFullYear();

  return day + " " + monthNames[monthIndex] + " " + year;
};

export const areDatesEqual = (date1, date2) => {
  return (
    date1.getFullYear() === date2.getFullYear() &&
    date1.getMonth() === date2.getMonth() &&
    date1.getDate() === date2.getDate()
  );
};
