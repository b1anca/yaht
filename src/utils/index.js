export const getRangeFromViewportWidth = () => {
  const viewportWidth = window.innerWidth;

  if (viewportWidth <= 390) {
    return 3;
  } else if (viewportWidth <= 500) {
    return 4;
    // sm
  } else if (viewportWidth <= 640) {
    return 6;
    // md
  } else if (viewportWidth <= 768) {
    return 8;
    // lg
  } else if (viewportWidth <= 1024) {
    return 9;
  } else {
    return 10;
  }
};
