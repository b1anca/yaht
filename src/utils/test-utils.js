import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { setupStore } from "../app/store";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  } = {}
) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  }
  return {
    store,
    user: userEvent.setup(),
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
