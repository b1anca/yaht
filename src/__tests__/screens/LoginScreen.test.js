import React from "react";
import { rest } from "msw";
import { fireEvent, screen } from "@testing-library/react";
import { renderWithProviders } from "../../utils/test-utils";
import { server } from "../../mocks/server";
import LoginScreen from "../../screens/LoginScreen";

describe("LoginScreen", () => {
  it("should render correctly", () => {
    const { asFragment } = renderWithProviders(<LoginScreen />);

    expect(asFragment()).toMatchSnapshot();
  });

  xit("should navigate to dashboard after form submit", async () => {
    const { user } = renderWithProviders(<LoginScreen />);
    const emailInput = screen.getByLabelText("email");
    const passwordInput = screen.getByLabelText("password");
    const email = "user@example.com";

    fireEvent.change(emailInput, { target: { value: email } });
    fireEvent.change(passwordInput, { target: { value: "examplePassword" } });
    await user.click(screen.getByRole("button", { name: "Sign in" }));

    screen.debug();
    expect(await screen.getByText(/dashboard/i)).toBeInTheDocument();

    // expect(navigate).toHaveBeenCalledWith("/dashboard");
    // expect(getByText(/dashboard/i)).toBeInTheDocument();
  });

  it("should be able to submit form and save token to local storage", async () => {
    const { user } = renderWithProviders(<LoginScreen />);
    const emailInput = screen.getByLabelText("email");
    const passwordInput = screen.getByLabelText("password");
    const email = "user@example.com";

    fireEvent.change(emailInput, { target: { value: email } });
    fireEvent.change(passwordInput, { target: { value: "examplePassword" } });
    await user.click(screen.getByRole("button", { name: "Sign in" }));

    expect(window.localStorage.getItem("userToken")).toEqual(`token-${email}`);
  });

  it("should show error message on server error", async () => {
    server.use(
      rest.post(
        `${process.env.REACT_APP_YAHT_API_URL}/auth/login`,
        (_, res, ctx) => {
          return res(ctx.json({ error: "unauthorized" }), ctx.status(401));
        }
      )
    );
    const { user } = renderWithProviders(<LoginScreen />);

    user.click(screen.getByRole("button", { name: "Sign in" }));
    await screen.findByRole("alert");

    expect(screen.getByRole("alert")).toHaveTextContent(
      "Request failed with status code 401"
    );
  });
});
