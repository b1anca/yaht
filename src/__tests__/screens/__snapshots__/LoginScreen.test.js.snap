// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen should render correctly 1`] = `
<DocumentFragment>
  <main
    class="relative h-screen flex flex-1 flex-col overflow-hidden px-4 py-8 sm:px-6 lg:px-8"
  >
    <div
      class="relative flex flex-1 flex-col items-center justify-center pb-16 pt-12"
    >
      <form
        class="w-full max-w-sm"
      >
        <div
          class="mb-6"
        >
          <label
            class="block text-sm font-semibold leading-6"
            for="email"
          >
            Email
          </label>
          <input
            aria-label="email"
            autocomplete="username"
            class="bg-transparent mt-2 appearance-none text-slate-100 rounded-md block w-full px-3 h-10 shadow-sm sm:text-sm focus:outline-none placeholder:text-slate-300 focus:ring-2 focus:ring-sky-500 ring-1 ring-slate-200"
            id="email"
            name="email"
            required=""
            type="email"
          />
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block text-sm font-semibold leading-6"
            for="password"
          >
            Password
          </label>
          <input
            aria-label="password"
            autocomplete="current-password"
            class="bg-transparent mt-2 appearance-none text-slate-100 rounded-md block w-full px-3 h-10 shadow-sm sm:text-sm focus:outline-none placeholder:text-slate-300 focus:ring-2 focus:ring-sky-500 ring-1 ring-slate-200"
            id="password"
            name="password"
            required=""
            type="password"
          />
        </div>
        <button
          class="text-sm button py-2 px-6 rounded-full inline-flex justify-center rounded-lg text-sm font-semibold py-2.5 px-4 bg-slate-300 text-slate-900 hover:bg-slate-200 w-full"
          type="submit"
        >
          Sign in
        </button>
      </form>
    </div>
    <footer
      class="relative shrink-0"
    >
      <div
        class="space-y-4 text-sm sm:flex sm:items-center sm:justify-center sm:space-x-4 sm:space-y-0"
      >
        <p
          class="text-center sm:text-left"
        >
          Don't have an account?
        </p>
        <a
          class="text-sm whitespace-nowrap py-2 px-6 rounded-full inline-flex justify-center rounded-lg font-semibold py-2.5 px-4 ring-1 ring-slate-300/10 hover:ring-slate-200/20"
          href="/register"
        >
          Register
        </a>
      </div>
    </footer>
  </main>
</DocumentFragment>
`;
